rfcs/1-input-pipeline.md

```diff
@@ -0,0 +1,391 @@
+## Background
+
+One of the primary functions of the Scikit-Learn wrappers for Keras is fielding conversion from Scikit-Learn's data types
+to TensorFlow's data types. An example of these conversions is integer-encoding or one-hot encoding categorical targets.
+
+Originally, these conversions were hardcoded into the wrappers in an ad-hoc manner (see here).
+SciKeras introduced the concept of `TargetTransformer` and `InputTransformer`,
+two Scikit-Learn style transformers that formalize this data conversion framework and allow users to insert their own custom pipeline.
+
+Currently, user-customization requires subclassing the wrappers,
+and composability is only provided via meta-transformers (i.e. a Scikit-Learn pipeline) of transformers.
+
+Seperately, SciKeras also implements data validations that mirrors what most Scikit-Learn estimators implement,
+for example to assert that `X` and `y` are of the same length or that `y` is purely numeric for regressors.
+SciKeras also validates and inspects the model, for example to make sure that the output shape matches the
+target's shape.
+These checks are helpful for simple models, but may be too restrictive for more complex scenarios, like
+multi-input/output models.
+
+This RFC proposes a unifified interface for composable data transformations and validations.
+The goal is to provide a pipeline of default transformations and validations that cover the simple use cases,
+while allowing users to easily remove checks or add transformation steps for more advanced use cases.
+
+Some of the speicific functional requirements are:
+1. Able to implement the default (i.e. current) data transformations and validations. This includes:
+ - Integer encoding targets for classifiers.
+ - One-hot encoding targets for classifiers using the categorical crossentropy loss.
+ - Converting class probability predictions into class predictions for classifiers.
+2. Able to implement user-defined transformations, including:
+ - Splitting the input and/or target into multiple inputs/outputs.
+ - Reshaping 2D inputs into 3D.
+3. Able to operate on array-like data (lists, Numpy arrays, Pandas DataFrames, etc.) as well as `tf.data.Dataset`s.
+4. Composable and modifiable without subclassing.
+
+## Proposal
+
+This proposal consists of 2 pipelines:
+1. A pipeline for preparing array-like data for a `tf.data.Dataset`. This would include, for example, integer-encoding object-dtype arrays (`tf.Tensors` can't hold objects).
+2. A pipeline for applying transformations to `tf.data.Dataset`s. For example, one-hot encoding targets for classifiers using the categorical crossentropy loss.
+
+If the data comes in as a `tf.data.Dataset`, the first pipeline is skipped. If not, it is run and the output is converted to a `Dataset`.
+The second pipeline is then run in all cases.
+
+These pipelines will consist of chained transformers implementing a Scikit-Learn-like interface, but without being restricted to the exact Scikit-Learn API.
```

Comment:
By "exact Scikit-Learn API" I mean roughly:
- having a `fit` method with the signature `fit(X, y=None) -> self`
- having a transform method with the signature `transform(X) -> X'`
- an `inverse_transform` method with the signature `inverse_transform(X') -> X`.
And no other methods (that are part of the API).
Also `inverse_transform(transform(X)) == X` is not a requirement of the API, but it certainly is the spirit of it (and how most sklearn transformers are implemented, when it makes sense to do so).

> And the proposed "Scikit-learn-like interface" is public, correct?

Yes, it would be public and consists of `ArrayTransformer` and `DatasetTransformer`.
These can be base classes, protocols/interfaces or just duck-typing (implementation detail).
The idea is that users could take our default data validation/transformations and mix/match with their own for use cases like multi-output models or multi-output class reweighting.

This proposed interface violates the sklearn interface in several ways:
- by passing `X`, `y` and `sample_weight` in `ArrayTransformer`
- by passing a `tf.data.Dataset` object in `DatasetTransformer`
- because these really aren't inverse transformations: the forward/input transformation transforms X & y, the output transformation transforms `y'` (y predictions) and/or `y` itself. So `inverse_transform(transform(X)) != X`.
- by allowing side effects to be applied to the `model` (this is an addition to the API, it doesn't strictly break the implementation, but it does break the spirit).

> How can this RFC be reworked to conform with the "exact Scikit-Learn API"?

I don't think it can be easily reworked: there are too many limitations in the sklearn transformer API ((and the authors have said as much themselves)[https://github.com/scikit-learn/scikit-learn/issues/3855#issuecomment-63213164]).

The good news is that the Scikit-Learn API is a strict subset of this API, so we could provide wrappers to convert the Scikit-Learn transformer API into this one (eg. by specifying what `X` is to your transformer and dispatching the methods).

But let me ask: why is it important to adhere to the Scikit-Learn API? I know there are good reasons, I just want understand which you are thinking of.
My thoughts are that if we can't interoperate directly (i.e. use `make_pipeline` and other facilities of sklearn), there is not much value in being semi-compatible. I also think there aren't that many things (beyond pipelines) that would be useful here. But this may be short-sighted, I'm open to other opinions.

Reply:
This library has one job: to bring the Scikit-learn API to Keras. That means there needs to be a ***really good*** reason to break the Scikit-learn API.

That reason needs to consider alternative implementations, including the one that exactly follows the Scikit-learn API. I imagine some questions that need to be answered include the following:


* What affect does the implementation have on developers?
* What affect does the implementation have on the user?
* What specific issues make the implementation inappropriate?